src/function/table/read_csv.cpp

```diff
@@ -44,6 +45,78 @@ void ReadCSVData::FinalizeRead(ClientContext &context) {
 	BaseCSVData::Finalize();
 }
 
+//! Function to do schema discovery over one CSV file or a list/glob of CSV files
+void SchemaDiscovery(ClientContext &context, ReadCSVData &result, CSVReaderOptions &options,
+                     vector<LogicalType> &return_types, vector<string> &names, MultiFileList &multi_file_list) {
+	vector<CSVSchema> schemas;
+	const auto option_og = options;
+
+	const auto file_paths = multi_file_list.GetAllFiles();
+
+	// Here what we want to do is to sniff a given number of lines, if we have many files, we might go through them
+	// to reach the number of lines.
+	const idx_t required_number_of_lines = options.sniff_size * options.sample_size_chunks;
+
+	idx_t total_number_of_rows = 0;
+	idx_t current_file = 0;
+	options.file_path = file_paths[current_file];
+
+	result.buffer_manager = make_shared_ptr<CSVBufferManager>(context, options, options.file_path, 0, false);
+	{
+		CSVSniffer sniffer(options, result.buffer_manager, CSVStateMachineCache::Get(context));
+		auto sniffer_result = sniffer.SniffCSV();
+		idx_t rows_read = sniffer.LinesSniffed() -
+		                  (options.dialect_options.skip_rows.GetValue() + options.dialect_options.header.GetValue());
+
+		schemas.emplace_back(sniffer_result.names, sniffer_result.return_types, file_paths[0], rows_read,
+		                     result.buffer_manager->GetBuffer(0)->actual_size == 0);
+		total_number_of_rows += sniffer.LinesSniffed();
+	}
+
+	// We do a copy of the options to not pollute the options of the first file.
+	while (total_number_of_rows < required_number_of_lines && current_file + 1 < file_paths.size()) {
```

Comment:
Should we add some upper-bound to how many files we read here, e.g. 10 files or so?